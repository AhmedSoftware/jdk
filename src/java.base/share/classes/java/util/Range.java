package java.util;

public sealed interface Range<T extends Comparable<T>> permits Range.BoundedAtEnd, Range.BoundedAtStart, Ranges.UnboundedRange {
    /**
     * Returns an unbounded range.
     *
     * @param <T> the type of the values
     * @return an unbounded range
     */
    static <T extends Comparable<T>> Range<T> unbounded() {
        return new Ranges.UnboundedRange<>();
    }

    /**
     * Returns an unbounded range ending at the given value.
     *
     * @param end the end of the range
     * @param <T> the type of the values
     * @return an unbounded range ending at the given value
     */
    static <T extends Comparable<T>> Range<T> unboundedEndingAt(T end) {
        return new Ranges.UnboundedStartRange<>(end);
    }

    /**
     * Returns an unbounded range ending at the given value.
     *
     * @param end        the end of the range
     * @param excludeEnd whether the end is exclusive
     * @param <T>        the type of the values
     * @return an unbounded range ending at the given value
     */
    static <T extends Comparable<T>> Range<T> unboundedEndingAt(T end, boolean excludeEnd) {
        return new Ranges.UnboundedStartRange<>(end, excludeEnd);
    }

    /**
     * Returns an unbounded range starting at the given value with the start included.
     *
     * @param start the start of the range
     * @param <T>   the type of the values
     * @return an unbounded range starting at the given value
     */
    static <T extends Comparable<T>> Range<T> unboundedStartingAt(T start) {
        return new Ranges.UnboundedEndRange<>(start);
    }

    /**
     * Returns an unbounded range starting at the given value.
     *
     * @param start        the start of the range
     * @param excludeStart whether the start is exclusive
     * @param <T>          the type of the values
     * @return an unbounded range starting at the given value
     */
    static <T extends Comparable<T>> Range<T> unboundedStartingAt(T start, boolean excludeStart) {
        return new Ranges.UnboundedEndRange<>(start, excludeStart);
    }

    /**
     * Returns a bounded range with both bounds included.
     *
     * @param start the start of the range
     * @param end   the end of the range
     * @param <T>   the type of the values
     * @return a bounded range
     */
    static <T extends Comparable<T>> Range<T> of(T start, T end) {
        return of(start, end, BoundsExclusion.NONE);
    }

    /**
     * Returns a bounded range with the given bounds exclusion.
     *
     * @param start           the start of the range
     * @param end             the end of the range
     * @param boundsExclusion the bounds exclusion
     * @param <T>             the type of the values
     * @return a bounded range with the given bounds exclusion
     */
    static <T extends Comparable<T>> Range<T> of(T start, T end, BoundsExclusion boundsExclusion) {
        if (start.compareTo(end) > 0) {
            throw new IllegalArgumentException("Start must be less than or equal to end");
        }
        return new Ranges.BoundedRange<>(start, end, boundsExclusion);
    }

    /**
     * Returns whether the span contains the given instant.
     *
     * @param instant the instant to check
     * @return true if the span contains the instant, false otherwise
     */
    boolean contains(T instant);

    /**
     * Returns whether the span overlaps with the given span.
     *
     * @param other the span to check
     * @return true if the span overlaps with the given span, false otherwise
     */
    boolean overlaps(Range<? extends T> other);

    /**
     * Returns the intersection of the span with the given span.
     *
     * @param other the span to intersect with
     * @return the intersection of the span with the given span. If the spans do not overlap, the result is an empty optional.
     */
    Optional<Range<? extends T>> intersection(Range<? extends T> other);

    /**
     * Returns the union of the span with the given span.
     *
     * @param other the span to union with
     * @return the union of the span with the given span.
     * If the spans do not overlap, the result is an array of two spans.
     * If the spans overlap, the result is an array of one combined span.
     */
    Ranges.Union<? extends T> union(Range<? extends T> other);

    /**
     * Returns the gap between the span and the given span.
     *
     * @param other the span to gap with
     * @return the gap between the span and the given span. If the spans overlap, the result is an empty optional.
     */
    Optional<Range<T>> gap(Range<? extends T> other);

    enum BoundsExclusion {
        NONE,
        START,
        END,
        BOTH
    }

    sealed interface BoundedAtEnd<T extends Comparable<T>> extends Range<T> permits Ranges.BoundedRange, Ranges.UnboundedStartRange {

        T end();

        boolean isEndExclusive();

        /**
         * Returns whether the span is before the given span.
         *
         * @param other the span to check
         * @return true if the span is before the given span, false otherwise
         */
        boolean isBefore(BoundedAtStart<? extends T> other);

        /**
         * Returns whether the span is before the given point.
         *
         * @param point the instant to check
         * @return true if the span is before the given instant, false otherwise
         */
        boolean isBefore(T point);
    }

    sealed interface BoundedAtStart<T extends Comparable<T>> extends Range<T> permits Ranges.BoundedRange, Ranges.UnboundedEndRange {

        T start();

        boolean isStartExclusive();

        /**
         * Returns whether the span is after the given span.
         *
         * @param other the span to check
         * @return true if the span is after the given span, false otherwise
         */
        boolean isAfter(BoundedAtEnd<? extends T> other);

        /**
         * Returns whether the span is after the given point.
         *
         * @param point the instant to check
         * @return true if the span is after the given instant, false otherwise
         */
        boolean isAfter(T point);
    }
}
