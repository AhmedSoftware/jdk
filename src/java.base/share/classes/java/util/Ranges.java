package java.util;

/**
 * Utility class containing range implementations.
 */
class Ranges {

    private static <T extends Comparable<T>> int endCompareOffset(Range.BoundedAtEnd<? extends T> other) {
        return other.isEndExclusive() ? 1 : 0;
    }

    private static <T extends Comparable<T>> int startCompareOffset(Range.BoundedAtStart<? extends T> other) {
        return other.isStartExclusive() ? 1 : 0;
    }

    private static <T extends Comparable<T>> boolean startBeforeOrAtEndOf(Range.BoundedAtStart<T> target, Range.BoundedAtEnd<? extends T> other) {
        return !startAfterEndOf(target, other);
    }

    private static <T extends Comparable<T>> boolean endAfterOrAtStartOf(Range.BoundedAtEnd<T> target, Range.BoundedAtStart<? extends T> other) {
        return !endBeforeStartOf(target, other);
    }

    private static <T extends Comparable<T>> boolean startBeforeOrAtStartOf(Range.BoundedAtStart<T> target, Range.BoundedAtStart<? extends T> other) {
        return !startAfterStartOf(target, other);
    }

    private static <T extends Comparable<T>> boolean endAfterOrAtEndOf(Range.BoundedAtEnd<T> target, Range.BoundedAtEnd<? extends T> other) {
        return !endBeforeEndOf(target, other);
    }

    private static <T extends Comparable<T>> boolean endBeforeOrAtStartOf(Range.BoundedAtEnd<T> target, Range.BoundedAtStart<? extends T> other) {
        return !endAfterStartOf(target, other);
    }

    private static <T extends Comparable<T>> boolean startAfterOrAtEndOf(Range.BoundedAtStart<T> target, Range.BoundedAtEnd<? extends T> other) {
        return !startBeforeEndOf(target, other);
    }

    private static <T extends Comparable<T>> boolean startAfterOrAtStartOf(Range.BoundedAtStart<T> target, Range.BoundedAtStart<? extends T> other) {
        return !startBeforeStartOf(target, other);
    }

    private static <T extends Comparable<T>> boolean endBeforeOrAtEndOf(Range.BoundedAtEnd<T> target, Range.BoundedAtEnd<? extends T> other) {
        return !endAfterEndOf(target, other);
    }

    @SuppressWarnings("unchecked")
    private static <T extends Comparable<T>> boolean endBeforeEndOf(Range.BoundedAtEnd<T> target, Range.BoundedAtEnd<? extends T> other) {
        return target.end().compareTo(other.end()) <= -(1 ^ endCompareOffset(target) | endCompareOffset(other));
    }

    @SuppressWarnings("unchecked")
    private static <T extends Comparable<T>> boolean startBeforeStartOf(Range.BoundedAtStart<T> target, Range.BoundedAtStart<? extends T> other) {
        return target.start().compareTo(other.start()) <= -(1 ^ startCompareOffset(target) | startCompareOffset(other));
    }

    @SuppressWarnings("unchecked")
    private static <T extends Comparable<T>> boolean startAfterStartOf(Range.BoundedAtStart<T> target, Range.BoundedAtStart<? extends T> other) {
        return target.start().compareTo(other.start()) >= (startCompareOffset(target) | (1 ^ startCompareOffset(other)));
    }

    @SuppressWarnings("unchecked")
    private static <T extends Comparable<T>> boolean endAfterEndOf(Range.BoundedAtEnd<T> target, Range.BoundedAtEnd<? extends T> other) {
        return target.end().compareTo(other.end()) >= (endCompareOffset(target) | (1 ^ endCompareOffset(other)));
    }

    @SuppressWarnings("unchecked")
    private static <T extends Comparable<T>> boolean startAfterEndOf(Range.BoundedAtStart<T> target, Range.BoundedAtEnd<? extends T> other) {
        return target.start().compareTo(other.end()) >= (1 ^ (startCompareOffset(target) | endCompareOffset(other)));
    }

    @SuppressWarnings("unchecked")
    private static <T extends Comparable<T>> boolean endBeforeStartOf(Range.BoundedAtEnd<T> target, Range.BoundedAtStart<? extends T> other) {
        return target.end().compareTo(other.start()) <= -(1 ^ (endCompareOffset(target) | startCompareOffset(other)));
    }

    @SuppressWarnings("unchecked")
    private static <T extends Comparable<T>> boolean startBeforeEndOf(Range.BoundedAtStart<T> target, Range.BoundedAtEnd<? extends T> other) {
        return target.start().compareTo(other.end()) < 0;
    }

    @SuppressWarnings("unchecked")
    private static <T extends Comparable<T>> boolean endAfterStartOf(Range.BoundedAtEnd<T> target, Range.BoundedAtStart<? extends T> other) {
        return target.end().compareTo(other.start()) > 0;
    }

    static final class UnboundedRange<T extends Comparable<T>> implements Range<T> {

        @Override
        public boolean contains(T instant) {
            return true;
        }

        @Override
        public boolean overlaps(Range<? extends T> other) {
            return true;
        }

        @Override
        public Optional<Range<? extends T>> intersection(Range<? extends T> other) {
            return Optional.of(other);
        }

        @Override
        public Union<T> union(Range<? extends T> other) {
            return new UnionOfOne<>(this);
        }

        @Override
        public Optional<Range<T>> gap(Range<? extends T> other) {
            return Optional.empty();
        }

        public boolean equals(Object obj) {
            return obj instanceof UnboundedRange;
        }

        public int hashCode() {
            return 0;
        }

        @Override
        public String toString() {
            return "-inf - inf";
        }
    }

    static final class UnboundedStartRange<T extends Comparable<T>> implements Range.BoundedAtEnd<T> {

        private final T end;
        private final boolean excludeEnd;

        public UnboundedStartRange(T end) {
            this(end, false);
        }

        public UnboundedStartRange(T end, boolean excludeEnd) {
            this.end = end;
            this.excludeEnd = excludeEnd;
        }

        @Override
        public T end() {
            return end;
        }

        @Override
        public boolean contains(T instant) {
            return instant.compareTo(end) <= -endCompareOffset(this);
        }

        @Override
        public boolean overlaps(Range<? extends T> other) {
            if (other instanceof Ranges.UnboundedRange) {
                return true;
            }
            if (other instanceof Range.BoundedAtStart<? extends T> otherStartBounded) {
                return endAfterOrAtStartOf(this, otherStartBounded);
            }
            return true;
        }

        @Override
        public boolean isBefore(Range.BoundedAtStart<? extends T> other) {
            return !overlaps(other);
        }

        @Override
        public boolean isBefore(T point) {
            return point.compareTo(end) <= 0;
        }

        @Override
        @SuppressWarnings("unchecked")
        public Optional<Range<? extends T>> intersection(Range<? extends T> other) {
            if (other instanceof Ranges.UnboundedRange) {
                return Optional.of(this);
            }
            if (other instanceof Range.BoundedAtStart<? extends T> otherBoundedAtStart) {
                if (endBeforeStartOf(this, otherBoundedAtStart)) {
                    return Optional.empty();
                }
                if (other instanceof Range.BoundedAtEnd<? extends T> otherEndBounded && endAfterOrAtEndOf(this, otherEndBounded)) {
                    // other is fully contained
                    return Optional.of(other);
                }
                return Optional.of(new BoundedRange<>(otherBoundedAtStart.start(), end));
            } else if (other instanceof Range.BoundedAtEnd<? extends T> otherEndBounded && endAfterOrAtEndOf(this, otherEndBounded)) {
                return Optional.of(other);
            }
            return Optional.of(this);
        }

        @Override
        @SuppressWarnings("unchecked")
        public Union<? extends T> union(Range<? extends T> other) {
            if (other instanceof Ranges.UnboundedRange) {
                return new UnionOfOne<T>(new UnboundedRange<>());
            }
            if (other instanceof Range.BoundedAtEnd<? extends T> otherEndBounded) {
                if (endBeforeOrAtEndOf(this, otherEndBounded)) {
                    return new UnionOfOne<>(new UnboundedStartRange<>(otherEndBounded.end()));
                }
            } else if (other instanceof Range.BoundedAtStart<? extends T> otherStartBounded) {
                if (endAfterOrAtStartOf(this, otherStartBounded)) {
                    return new UnionOfOne<T>(new UnboundedRange<>());
                }
            }
            return new UnionOfTwo<>(this, (Range<T>) other);
        }

        @Override
        @SuppressWarnings("unchecked")
        public Optional<Range<T>> gap(Range<? extends T> other) {
            if (other instanceof Ranges.UnboundedRange) {
                return Optional.empty();
            }
            if (other instanceof Range.BoundedAtStart<? extends T> otherStartBounded) {
                var compResult = end.compareTo((T) otherStartBounded.start());
                if (compResult > 0 || (compResult == 0 && excludeEnd && otherStartBounded.isStartExclusive())) {
                    return Optional.of(new BoundedRange<>(end, otherStartBounded.start()));
                }
            }
            return Optional.empty();
        }

        @Override
        public boolean isEndExclusive() {
            return excludeEnd;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (!(o instanceof UnboundedStartRange<?> that)) {
                return false;
            }
            return Objects.equals(end, that.end);
        }

        @Override
        public int hashCode() {
            return Objects.hashCode(end);
        }

        @Override
        public String toString() {
            return "-inf - " + end;
        }
    }

    static final class UnboundedEndRange<T extends Comparable<T>> implements Range.BoundedAtStart<T> {

        private final T start;

        private final boolean excludeStart;

        public UnboundedEndRange(T start) {
            this(start, false);
        }

        public UnboundedEndRange(T start, boolean excludeStart) {
            this.start = start;
            this.excludeStart = excludeStart;
        }

        @Override
        public T start() {
            return start;
        }

        @Override
        public boolean contains(T instant) {
            return instant.compareTo(start) >= (startCompareOffset(this));
        }

        @Override
        public boolean overlaps(Range<? extends T> other) {
            if (other instanceof Ranges.UnboundedRange) {
                return true;
            }
            if (other instanceof Range.BoundedAtEnd<? extends T> otherEndBounded) {
                return startBeforeOrAtEndOf(this, otherEndBounded);
            }
            return true;
        }

        @Override
        public boolean isAfter(Range.BoundedAtEnd<? extends T> other) {
            return !overlaps(other);
        }

        @Override
        public boolean isAfter(T point) {
            return start.compareTo(point) < 0;
        }

        @Override
        @SuppressWarnings("unchecked")
        public Optional<Range<? extends T>> intersection(Range<? extends T> other) {
            if (other instanceof Range.BoundedAtEnd<? extends T> otherEndBounded) {
                if (other instanceof Range.BoundedAtStart<? extends T> otherStartBounded) {
                    if (startAfterEndOf(this, otherEndBounded)) {
                        return Optional.empty();
                    } else if (startBeforeOrAtStartOf(this, otherStartBounded)) {
                        return Optional.of(other);
                    }
                } else if (startAfterEndOf(this, otherEndBounded)) {
                    return Optional.empty();
                }
                return Optional.of(new BoundedRange<>(start, otherEndBounded.end()));
            } else if (other instanceof Range.BoundedAtStart<? extends T> otherStartBounded) {
                if (startBeforeOrAtStartOf(this, otherStartBounded)) {
                    return Optional.of(other);
                }
            }
            return Optional.of(this);
        }

        @Override
        @SuppressWarnings("unchecked")
        public Union<? extends T> union(Range<? extends T> other) {
            if (other instanceof Range.BoundedAtEnd<? extends T> otherEndBounded) {
                if (other instanceof Range.BoundedAtStart<? extends T> otherStartBounded) {
                    if (startBeforeOrAtEndOf(this, otherEndBounded)) {
                        if (startAfterOrAtStartOf(this, otherStartBounded)) {
                            return new UnionOfOne<>(new UnboundedEndRange<>(otherStartBounded.start()));
                        }
                        return new UnionOfOne<>(this);
                    }
                } else if (startBeforeOrAtEndOf(this, otherEndBounded)) {
                    return new UnionOfOne<>(new UnboundedRange<T>());
                }
            } else if (other instanceof Range.BoundedAtStart<? extends T> otherStartBounded) {
                return startAfterStartOf(this, otherStartBounded) ? new UnionOfOne<>((Range<T>) other) : new UnionOfOne<>(this);
            }
            return new UnionOfTwo<>(this, (Range<T>) other);
        }

        @Override
        @SuppressWarnings("unchecked")
        public Optional<Range<T>> gap(Range<? extends T> other) {
            if (other instanceof Ranges.UnboundedRange) {
                return Optional.empty();
            }
            if (other instanceof Range.BoundedAtEnd<? extends T> otherEndBounded) {
                var compResult = start.compareTo((T) otherEndBounded.end());
                if (compResult < 0 || (compResult == 0 && excludeStart && otherEndBounded.isEndExclusive())) {
                    return Optional.of(new BoundedRange<>(otherEndBounded.end(), start));
                }
            }
            return Optional.empty();
        }

        @Override
        public boolean isStartExclusive() {
            return excludeStart;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (!(o instanceof UnboundedEndRange<?> that)) {
                return false;
            }
            return Objects.equals(start, that.start);
        }

        @Override
        public int hashCode() {
            return Objects.hashCode(start);
        }

        @Override
        public String toString() {
            return start + " - inf";
        }
    }

    static final class BoundedRange<T extends Comparable<T>> implements Range.BoundedAtStart<T>, Range.BoundedAtEnd<T> {

        private final T start;
        private final T end;
        private final boolean excludeStart;
        private final boolean excludeEnd;

        public BoundedRange(T start, T end) {
            this(start, end, BoundsExclusion.NONE);
        }

        public BoundedRange(T start, T end, BoundsExclusion boundsExclusion) {
            this.start = start;
            this.end = end;
            this.excludeStart = boundsExclusion == BoundsExclusion.START || boundsExclusion == BoundsExclusion.BOTH;
            this.excludeEnd = boundsExclusion == BoundsExclusion.END || boundsExclusion == BoundsExclusion.BOTH;
        }

        @Override
        public T start() {
            return start;
        }

        @Override
        public T end() {
            return end;
        }

        @Override
        public boolean contains(T instant) {
            return instant.compareTo(start) >= 0 && instant.compareTo(end) <= 0;
        }

        @Override
        public boolean overlaps(Range<? extends T> other) {
            if (other instanceof Range.BoundedAtEnd<? extends T> otherEndBounded) {
                if (other instanceof Range.BoundedAtStart<? extends T> otherStartBounded) {
                    return endAfterOrAtStartOf(this, otherStartBounded) && startBeforeOrAtEndOf(this, otherEndBounded);
                }
                return startBeforeOrAtEndOf(this, otherEndBounded);
            } else if (other instanceof Range.BoundedAtStart<? extends T> otherStartBounded) {
                return endAfterOrAtStartOf(this, otherStartBounded);
            }
            return true;
        }

        @Override
        public boolean isBefore(T point) {
            return point.compareTo(end) <= 0;
        }

        @Override
        public boolean isBefore(Range.BoundedAtStart<? extends T> other) {
            return endBeforeStartOf(this, other);
        }

        @Override
        public boolean isAfter(Range.BoundedAtEnd<? extends T> other) {
            return startAfterEndOf(this, other);
        }

        @Override
        public boolean isAfter(T point) {
            return start.compareTo(point) > 0;
        }

        @Override
        @SuppressWarnings("unchecked")
        public Optional<Range<? extends T>> intersection(Range<? extends T> other) {
            if (other instanceof Range.BoundedAtStart<? extends T> otherStartBounded) {
                if (other instanceof Range.BoundedAtEnd<? extends T> otherEndBounded) {
                    if (endBeforeStartOf(this, otherStartBounded) || startAfterEndOf(this, otherEndBounded)) {
                        return Optional.empty();
                    } else if (endAfterOrAtEndOf(this, otherEndBounded)) {
                        if (startBeforeOrAtStartOf(this, otherStartBounded)) {
                            return Optional.of(other);
                        } else if (startBeforeEndOf(this, otherEndBounded)) {
                            return Optional.of(new BoundedRange<>(start, otherEndBounded.end()));
                        } else {
                            return Optional.empty();
                        }
                    } else {
                        if (startBeforeOrAtStartOf(this, otherStartBounded)) {
                            return Optional.of(new BoundedRange<>(otherStartBounded.start(), end));
                        }
                        return Optional.of(this);
                    }
                } else if (endBeforeStartOf(this, otherStartBounded)) {
                    return Optional.empty();
                }
            } else if (other instanceof Range.BoundedAtEnd<? extends T> otherEndBounded) {
                if (startAfterEndOf(this, otherEndBounded)) {
                    return Optional.empty();
                }
            }
            return Optional.of(this);
        }

        @Override
        @SuppressWarnings("unchecked")
        public Union<? extends T> union(Range<? extends T> other) {
            if (other instanceof Range.BoundedAtStart<? extends T> otherStartBounded) {
                if (other instanceof Range.BoundedAtEnd<? extends T> && endAfterOrAtStartOf(this, otherStartBounded)) {
                    if (otherStartBounded.start().compareTo(start) >= 0) {
                        return new UnionOfOne<>((Range<T>) other);
                    }
                    return new UnionOfOne<>(new BoundedRange<>(otherStartBounded.start(), end));
                } else {
                    if (startAfterOrAtStartOf(this, otherStartBounded)) {
                        return new UnionOfOne<>((Range<T>) other);
                    } else if (otherStartBounded.start().compareTo(end) <= 0) {
                        return new UnionOfOne<>(new UnboundedEndRange<>(otherStartBounded.start()));
                    }
                }
            } else if (other instanceof Range.BoundedAtEnd<? extends T> otherEndBounded) {
                if (endBeforeOrAtEndOf(this, otherEndBounded)) {
                    return new UnionOfOne<>((Range<T>) other);
                } else if (startBeforeOrAtEndOf(this, otherEndBounded)) {
                    return new UnionOfOne<>(new UnboundedStartRange<>(otherEndBounded.end()));
                }
            }
            return new UnionOfTwo<>(this, (Range<T>) other);
        }

        @Override
        @SuppressWarnings("unchecked")
        public Optional<Range<T>> gap(Range<? extends T> other) {
            if (other instanceof Range.BoundedAtStart<? extends T> otherStartBounded && endBeforeStartOf(this, otherStartBounded)) {
                return Optional.of(new BoundedRange<>(end, otherStartBounded.start()));
            } else if (other instanceof Range.BoundedAtEnd<? extends T> otherEndBounded && startAfterEndOf(this, otherEndBounded)) {
                return Optional.of(new BoundedRange<>(otherEndBounded.end(), start));
            }
            return Optional.empty();
        }

        @Override
        public boolean isEndExclusive() {
            return excludeEnd;
        }

        @Override
        public boolean isStartExclusive() {
            return excludeStart;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (!(o instanceof BoundedRange<?> that)) {
                return false;
            }
            return Objects.equals(start, that.start) && Objects.equals(end, that.end);
        }

        @Override
        public int hashCode() {
            return Objects.hash(start, end);
        }

        @Override
        public String toString() {
            return start + " - " + end;
        }
    }

    /**
     * Represents a union of two spans.
     * The spans are guaranteed to either abut or not touch at all.
     *
     * @param <T> the type of the values
     */
    public sealed interface Union<T extends Comparable<T>> {
    }

    /**
     * Represents a result of union operation where the result is a single span.
     *
     * @param <T> the type of the values
     */
    public static final class UnionOfOne<T extends Comparable<T>> implements Union<T> {

        private final Range<T> range;

        private UnionOfOne(Range<T> range) {
            this.range = range;
        }

        /**
         * Returns the span.
         *
         * @return the span
         */
        public Range<T> range() {
            return range;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (!(o instanceof UnionOfOne<?> that)) {
                return false;
            }
            return Objects.equals(range, that.range);
        }
    }

    /**
     * Represents a result of union operation where the result is two spans.
     *
     * @param <T> the type of the values
     */
    public static final class UnionOfTwo<T extends Comparable<T>> implements Union<T> {

        private final Range<T> first;
        private final Range<T> second;

        private UnionOfTwo(Range<T> first, Range<T> second) {
            this.first = first;
            this.second = second;
        }

        /**
         * Returns the first span.
         *
         * @return the first span
         */
        public Range<T> first() {
            return first;
        }

        /**
         * Returns the second span.
         *
         * @return the second span
         */
        public Range<T> second() {
            return second;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (!(o instanceof UnionOfTwo<?> that)) {
                return false;
            }
            return Objects.equals(first, that.first) && Objects.equals(second, that.second);
        }
    }

}