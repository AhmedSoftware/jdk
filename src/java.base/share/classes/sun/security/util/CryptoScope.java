package sun.security.util;

import java.security.CryptoPrimitive;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Crypto scope corresponds to one or more crypto primitives.
 */
enum CryptoScope {

    KX("kx", new HashSet<>(List.of(CryptoPrimitive.KEY_AGREEMENT))),      // Key Exchange
    AUTHN("authn", new HashSet<>(List.of(CryptoPrimitive.SIGNATURE)));       // Authentication

    private final Set<CryptoPrimitive> cryptoPrimitives;
    private final String name;

    CryptoScope(String name, Set<CryptoPrimitive> cryptoPrimitives) {
        this.name = name;
        this.cryptoPrimitives = cryptoPrimitives;
    }

    Set<CryptoPrimitive> getCryptoPrimitives() {
        return cryptoPrimitives;
    }

    String getName() {
        return name;
    }
}
