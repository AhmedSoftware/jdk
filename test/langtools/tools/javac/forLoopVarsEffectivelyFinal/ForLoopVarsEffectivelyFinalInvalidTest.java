/*
 * @test /nodynamiccopyright/
 * @bug 8338711
 * @summary Check for invalid cases relating to for() loop variables being effectively final in the body
 * @compile/fail/ref=ForLoopVarsEffectivelyFinalInvalidTest.out -XDrawDiagnostics ForLoopVarsEffectivelyFinalInvalidTest.java
 */

import java.util.function.*;

public class ForLoopVarsEffectivelyFinalInvalidTest {

    // If the variable is mutated in the BODY, it loses its effectively final status in the BODY
    {
        for (int i = 0; i < 3; ) {
            i++;
            ((IntSupplier)() -> f(i)).getAsInt();                       // error: ...must be final or effectively final
        }
    }

    // If the variable is mutated in the INIT, it loses its effectively final status in the INIT
    {
        for (int i = 0, j = i++ + ((IntSupplier)() -> f(i)).getAsInt(); // error: ...must be final or effectively final
            i < 3; ) {
            ((IntSupplier)() -> f(i)).getAsInt();                       // ok
        }
    }

    // If the variable is mutated in the CONDITION, it loses its effectively final status in the CONDITION
    {
        for (int i = 0;
          i++ + ((IntSupplier)() -> f(i)).getAsInt() < 3; ) {           // error: ...must be final or effectively final
            ((IntSupplier)() -> f(i)).getAsInt();                       // ok
        }
    }

    // If the variable is mutated in the STEP, it loses its effectively final status in the STEP
    {
        for (int i = 0;
          i < 3;
          i += ((IntSupplier)() -> f(i)).getAsInt()) {                  // error: ...must be final or effectively final
            ((IntSupplier)() -> f(i)).getAsInt();                       // ok
        }
    }

    // An extra level of nesting doesn't make it OK
    {
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                ((IntSupplier)() -> f(i)).getAsInt();                   // error: ...must be final or effectively final
            }
            i++;
        }
    }

    // If the variable is never initialized, it can't be captured
    {
        for (int i; "".hashCode() == 0; ) {
            ((IntSupplier)() -> f(i)).getAsInt();                       // error: variable i might not have been initialized
        }
    }

    // An extra level of nesting doesn't make it OK
    {
        for (int i = 0; true; ) {
            for (int j = 0; j < 3; j++) {
                i = 42;
                Runnable r = () -> System.out.println(i);               // error: ...must be final or effectively final
            }
            break;
        }
    }

    // An extra level of nesting doesn't make it OK
    {
        for (int i = 0; true; ) {
            for (int j = 0; j < ++i; j++) {
                Runnable r = () -> System.out.println(i);               // error: ...must be final or effectively final
            }
            break;
        }
    }

    int f(int i) {
        return i * i;
    }
}
