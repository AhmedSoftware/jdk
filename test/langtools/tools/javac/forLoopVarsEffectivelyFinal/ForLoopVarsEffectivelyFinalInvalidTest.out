ForLoopVarsEffectivelyFinalInvalidTest.java:58:35: compiler.err.var.might.not.have.been.initialized: i
ForLoopVarsEffectivelyFinalInvalidTest.java:16:35: compiler.err.cant.ref.non.effectively.final.var: i, (compiler.misc.lambda)
ForLoopVarsEffectivelyFinalInvalidTest.java:22:57: compiler.err.cant.ref.non.effectively.final.var: i, (compiler.misc.lambda)
ForLoopVarsEffectivelyFinalInvalidTest.java:31:39: compiler.err.cant.ref.non.effectively.final.var: i, (compiler.misc.lambda)
ForLoopVarsEffectivelyFinalInvalidTest.java:40:38: compiler.err.cant.ref.non.effectively.final.var: i, (compiler.misc.lambda)
ForLoopVarsEffectivelyFinalInvalidTest.java:49:39: compiler.err.cant.ref.non.effectively.final.var: i, (compiler.misc.lambda)
ForLoopVarsEffectivelyFinalInvalidTest.java:67:55: compiler.err.cant.ref.non.effectively.final.var: i, (compiler.misc.lambda)
ForLoopVarsEffectivelyFinalInvalidTest.java:77:55: compiler.err.cant.ref.non.effectively.final.var: i, (compiler.misc.lambda)
8 errors
